python/dgl/dataloading/dataloader.py

```diff
@@ -133,6 +132,61 @@ def _get_id_tensor_from_mapping(indices, device, keys):
     return id_tensor
 
 
+def _split_to_local_id_tensor_from_mapping(indices, keys, bounds):
+    dtype = dtype_of(indices)
+    device = next(iter(indices.values())).device
+    num_samples = bounds[1] - bounds[0]
+    id_tensor = torch.empty(num_samples, 2, dtype=dtype, device=device)
+
+    index_offset = 0
+    split_id_offset = 0
+    for i, k in enumerate(keys):
+        if k not in indices:
+            continue
+        index = indices[k]
+        length = index.shape[0]
+        index_offset2 = index_offset + length
+        lower = max(bounds[0], index_offset)
+        upper = min(bounds[1], index_offset2)
+        if upper > lower:
+            split_id_offset2 = split_id_offset + (upper - lower)
+            assert split_id_offset2 <= num_samples
+            id_tensor[split_id_offset:split_id_offset2, 0] = i
+            id_tensor[split_id_offset:split_id_offset2, 1] = index[
+                lower - index_offset : upper - index_offset
+            ]
+            split_id_offset += upper - lower
+            if split_id_offset2 == num_samples:
+                break
+        index_offset = index_offset2
+
+    # padding locally when drop_last = False
+    if split_id_offset2 < num_samples:
+        padding_size = bounds[1] - index_offset2
+        id_tensor[split_id_offset2:, 0] = id_tensor[:padding_size, 0]
+        id_tensor[split_id_offset2:, 1] = id_tensor[:padding_size, 1]
+        assert num_samples == split_id_offset2 + padding_size
+    return id_tensor
+
+
+def _split_to_local_id_tensor(indices, bounds):
+    dtype = dtype_of(indices)
+    device = indices.device
+    num_samples = bounds[1] - bounds[0]
+    id_tensor = torch.empty(num_samples, dtype=dtype, device=device)
+
+    # if needs padding when drop_last = False
+    if bounds[1] > len(indices):
+        reminder = len(indices) - bounds[0]
+        id_tensor[0:reminder] = indices[bounds[0] :]
+        padding_size = bounds[1] - len(indices)
+        # padding id_tensor from start
```

Comment:
Do we need to pad `self._id_tensor`?
I thought we did not need to since at line 321, we make sure to pad `self._indices` with indices less than `self.local_num_indices`.

Reply:
Right. `self._id_tensor` may not need to be padded. I guess I was trying to keep the same size of `id_tensor` at every proc and did the padding for safety... I will simplify this part.

Reply:
My understanding is we need padding here for the case where the dataset is not shuffled, and `321` for when the dataset is shuffled (to ensure the replicated indices change with each shuffle).

Reply:
Ah the padding here is more for the non-shuffling case? But we could have the same behavior as un the shuffling case where we limit the index range inside the padded `self._indices` to be less than `self.local_num_indices` so that the padded section of `_id_tensor` is never used.

Reply:
Yes, we need to either pad `self._indices` or `self._id_tensor` for shuffling. In general, it doesn't matter as it would not affect memory and computation efficiency...

In the latest commit, I padded `self._indices` and leave `self._id_tensor` uninitialized, to make it consistent between `shuffle=True` or `False`